hack/lib/golang.sh

```diff
@@ -466,13 +466,23 @@ kube::golang::create_gopath_tree() {
 kube::golang::verify_go_version() {
   # default GO_VERSION to content of .go-version
   GO_VERSION="${GO_VERSION:-"$(cat "${KUBE_ROOT}/.go-version")"}"
-  # only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match GO_VERSION
-  if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
-      (command -v go >/dev/null && [ "$(go version | cut -d' ' -f3)" = "go${GO_VERSION}" ])); then
+  if [ "${GOTOOLCHAIN:-auto}" != 'auto' ]; then
+    # no-op, just respect GOTOOLCHAIN
+    :
+  elif [ -n "${FORCE_HOST_GO:-}" ]; then
+    # ensure existing host version is used, like before GOTOOLCHAIN existed
+    export GOTOOLCHAIN='local'
+  else
+    # otherwise, we want to ensure the go version matches GO_VERSION
+    GOTOOLCHAIN="go${GO_VERSION}"
+    export GOTOOLCHAIN
```

Comment:
I think this should move inside the if block where we don't have the go version we want... I'm seeing behavior where setting `GOTOOLCHAIN=go1.21.1` explicitly is making `go1.21.1` try to redownload (maybe related to build modifiers like boringcrypto be enabled)

Reply:
`go env GOVERSION` reports `go1.21.1 X:boringcrypto` when boringcrypto is enabled. The `go version` check we're doing below ignores any experiments, so when we force GOTOOLCHAIN to exactly `go1.21.1`, we might be creating the mismatch.

Worth filing an issue for the go team to clarify how we should be using GOTOOLCHAIN in a situation like this.

Reply:
ACK, I suspect `export GOTOOLCHAIN=local` is the right answer when using a toolchain like this.
If not I think we could add a check for `X:` experiments early in the version check and assume that we should behave like `GOTOOLCHAIN=local` if an experimental toolchain is being used.

Reply:
> ACK, I suspect export GOTOOLCHAIN=local is the right answer when using a toolchain like this

I don't think so ... `GOTOOLCHAIN=local` won't do the right thing when we propagate into a build container with a different go version (like our kube-cross images)

Reply:
[we've reached out for clarification with the go team and will follow-up...]

Reply:
forgot to circle back here -- The expectation is that users will use GOEXPERIMENT which works with GOTOOLCHAIN (and this PR), GOTOOLCHAIN will download the standard toolchain but you can use the standard toolchain with GOEXPERIMENT instead of a modified toolchain, and that issue generally exists for 1.21+ in all projects.